docs/design/proposals/cephfs-snapshot-shallow-ro-vol.md

```diff
@@ -0,0 +1,274 @@
+# Snapshots as shallow read-only volumes
+
+CSI spec doesn't have a notion of "mounting a snapshot". Instead, the idiomatic
+way of accessing snapshot contents is first to create a volume populated with
+snapshot contents and then mount that volume to workloads.
+
+CephFS exposes snapshots as special, read-only directories of a subvolume
+located in `<subvolume>/.snap`. cephfs-csi can already provision writable
+volumes with snapshots as their data source, where snapshot contents are
+cloned to the newly created volume. However, cloning a snapshot to volume
+is a very expensive operation in CephFS as the data needs to be fully copied.
+When the need is to only read snapshot contents, snapshot cloning is extremely
+inefficient and wasteful.
+
+This proposal describes a way for cephfs-csi to expose CephFS snapshots
+as shallow, read-only volumes, without needing to clone the underlying
+snapshot data.
+
+## Use-cases
+
+What's the point of such read-only volumes?
+
+* **Restore snapshots selectively:** users may want to traverse snapshots,
+  restoring data to a writable volume more selectively instead of restoring
+  the whole snapshot.
+* **Volume backup:** users can't backup a live volume, they first need
+  to snapshot it. Once a snapshot is take, it still can't be backed-up,
+  as backup tools usually work with volumes (that are exposed as file-systems)
+  and not snapshots (which might have backend-specific format). What this means
+  is that in order to create a snapshot backup, users have to clone snapshot
+  data twice:
+
+  1. first time, when restoring the snapshot into a temporary volume from
+     where the data will be read,
+  1. and second time, when transferring that volume into some backup/archive
+     storage (e.g. object store).
+
+  The temporary backed-up volume will most likely be thrown away after the
+  backup transfer is finished. That's a lot of wasted work for what we
+  originally wanted to do! Having the ability to create volumes from
+  snapshots cheaply would be a big improvement for this use case.
+
+## Alternatives
+
+* _Snapshots are stored in `<subvolume>/.snap`. Users could simply visit this
+  directory by themselves._
+
+  `.snap` is CephFS-specific detail of how snapshots are exposed.
+  Users / tools may not be aware of this special directory, or it may not fit
+  their workflow. At the moment, the idiomatic way of accessing snapshot
+  contents in CSI drivers is by creating a new volume and populating it
+  with snapshot.
+
+## Design
+
+Key points:
+
+* Volume source is a snapshot, volume access mode is `*_READER_ONLY`.
```

Comment:
This `*_READER ONLY` is bit complex or an area at kubernetes where things are vague or incomplete. However there are recent improvements on this , but yet to be completed. so I am not sure how, is it a good idea to rely on READER ONLY access mode. In other side, considering this is a design proposal may be we are good and later while coming into implementation we can check thoroughly.

Reply:
I didn't want to pollute storage class parameters any more than necessary. I don't have a test implementation of this yet so I'm not sure 100% either about this. If you're ok with this for now, we can see during implementation, like you've suggested.